lib/sparse/src/index/search_context.rs

```diff
@@ -140,27 +135,32 @@ impl<'a> SearchContext<'a> {
     /// b,  30, 34, 60, 230
     fn advance(&mut self) -> Option<ScoredPointOffset> {
         // Get current min record id from all posting list iterators
-        let current_min_record_id = *self.min_record_ids.peek()?;
+        let current_min_record_id = self.min_record_id?;
         let mut score = 0.0;
-
+        let mut found = false;
         // Iterate to advance matching posting iterators
         for posting_iterator in self.postings_iterators.iter_mut() {
             if let Some(element) = posting_iterator.posting_list_iterator.peek() {
                 // accumulate score for the current record id
                 if element.record_id == current_min_record_id {
+                    found = true;
                     score += element.weight * posting_iterator.query_weight;
                     // advance posting list iterator to next element
                     posting_iterator.posting_list_iterator.advance();
-                    // pop min record id
-                    self.min_record_ids.pop();
-                    // look-ahead to find next min record id
-                    if let Some(next_element) = posting_iterator.posting_list_iterator.peek() {
-                        self.min_record_ids.push(next_element.record_id);
-                    }
                 }
             }
         }
 
+        // update min record id for next iteration
+        if found {
+            // assume the next min record id is the current one + 1
+            self.min_record_id = self.min_record_id.map(|min_id| min_id + 1);
+        } else {
+            // no match found, compute next min record id from all posting list iterators
+            self.min_record_id = Self::next_min_id(&self.postings_iterators);
+            return self.advance();
```

Comment:
the recursion is bounded as the nested call will always return right away because it is using the proper min record id